File: JavaProject/GrowNet/src/main/java/ai/nektron/grownet/Layer.java
package ai.nektron.grownet;

import java.util.*;

/**
 * A mixed‑type population (excitatory, inhibitory, modulatory) that shares a lateral bus.
 */
public class Layer {

    private final List<Neuron> neurons = new ArrayList<>();
    private final LateralBus bus = new LateralBus();
    private final Random rng = new Random(1234);

    public Layer(int excitatoryCount, int inhibitoryCount, int modulatoryCount) {
        // Default slot policy for this demo layer.
        SlotConfig cfg = SlotConfig.fixed(10.0);  // 10% Δ bins
        int slotLimit = -1;                               // unlimited

        for (int i = 0; i < excitatoryCount; ++i) {
            Neuron n = new ExcitatoryNeuron("E" + i, bus, cfg, slotLimit);
            neurons.add(n);
        }
        for (int i = 0; i < inhibitoryCount; ++i) {
            Neuron n = new InhibitoryNeuron("I" + i, bus, cfg, slotLimit);
            neurons.add(n);
        }
        for (int i = 0; i < modulatoryCount; ++i) {
            Neuron n = new ModulatoryNeuron("M" + i, bus, cfg, slotLimit);
            neurons.add(n);
        }
    }

    // --------------------------------------------- accessors ------------------------------------------------

    public List<Neuron> neurons() { return neurons; }

    // ------------------------------------------ intra‑layer wiring ------------------------------------------

    /** Layer‑local fanout (demo‑level). */
    public void wireRandomFeedforward(double probability) {
        for (Neuron a : neurons) {
            for (Neuron b : neurons) {
                if (a == b) continue;
                if (rng.nextDouble() < probability) {
                    a.connect(b, /*feedback=*/false);
                }
            }
        }
    }

    public void wireRandomFeedback(double probability) {
        for (Neuron a : neurons) {
            for (Neuron b : neurons) {
                if (a == b) continue;
                if (rng.nextDouble() < probability) {
                    a.connect(b, /*feedback=*/true);
                }
            }
        }
    }

    // ------------------------------------------------ forward ------------------------------------------------

    /** Drive all neurons with a scalar value for this tick. */
    public void forward(double value) {
        for (Neuron n : neurons) {
            boolean fired = n.onInput(value);
            if (fired) {
                n.onOutput(value);
            }
        }
    }

    /** End‑of‑tick housekeeping: decay inhibition/modulation back toward neutral. */
    public void endTick() {
        bus.decay();
    }
}

File: grownet_python_mojo/python/layer.py
from typing import List
from region_bus import RegionBus
from neuron import Neuron, ExcitatoryNeuron, InhibitoryNeuron, ModulatoryNeuron

class Layer:
    def __init__(self, excitatory_count: int, inhibitory_count: int, modulatory_count: int, bus: RegionBus) -> None:
        self._bus = bus
        self._neurons: List[Neuron] = []
        for i in range(excitatory_count):
            self._neurons.append(ExcitatoryNeuron(f"E{i}", bus))
        for i in range(inhibitory_count):
            self._neurons.append(InhibitoryNeuron(f"I{i}", bus))
        for i in range(modulatory_count):
            self._neurons.append(ModulatoryNeuron(f"M{i}", bus))

    def get_neurons(self) -> List[Neuron]:
        return self._neurons

    def forward(self, value: float) -> None:
        for n in self._neurons:
            n.on_input(value)

    def end_tick(self) -> None:
        # bus maintenance + neuron maintenance
        self._bus.end_tick()
        for n in self._neurons:
            n.end_tick()
